src/init/src/main.rs

```diff
@@ -0,0 +1,85 @@
+mod sys;
+use sys::{freopen, mount, reboot};
+
+mod dmesg;
+use dmesg::{println, eprintln};
+
+mod platforms;
+
+// TODO: call seed_entropy with a generic source on non-aws targets, maybe by providing get_entropy
+// as an Option<fn> rather than fn.
+#[cfg(feature = "platform-aws")]
+use sys::seed_entropy;
+
+#[cfg(feature = "platform-aws")]
+use platforms::aws::{get_entropy, init_platform};
+
+/// Mount common filesystems with conservative permissions.
+fn init_rootfs() {
+    use libc::{MS_NODEV, MS_NOEXEC, MS_NOSUID};
+    let no_dse = MS_NODEV | MS_NOSUID | MS_NOEXEC;
+    let no_se = MS_NOSUID | MS_NOEXEC;
+    let args = [
+        ("devtmpfs", "/dev", "devtmpfs", no_se, "mode=0755"),
+        ("devtmpfs", "/dev", "devtmpfs", no_se, "mode=0755"),
+        ("devpts", "/dev/pts", "devpts", no_se, ""),
+        ("shm", "/dev/shm", "tmpfs", no_dse, "mode=0755"),
+        ("proc", "/proc", "proc", no_dse, "hidepid=2"),
+        ("tmpfs", "/run", "tmpfs", no_dse, "mode=0755"),
+        ("tmpfs", "/tmp", "tmpfs", no_dse, ""),
+        ("sysfs", "/sys", "sysfs", no_dse, ""),
+        (
+            "cgroup_root",
+            "/sys/fs/cgroup",
+            "tmpfs",
+            no_dse,
+            "mode=0755",
+        ),
+    ];
+    for (src, target, fstype, flags, data) in args {
+        match mount(src, target, fstype, flags, data) {
+            Ok(()) => println!("Mounted {target}"),
+            Err(e) => eprintln!("Unable to mount {target} ({e})"),
+        }
+    }
+}
+
+/// Initialize console with stdin/stdout/stderr.
+fn init_console() {
+    let args = [
+        ("/dev/console", "r", 0),
+        ("/dev/console", "w", 1),
+        ("/dev/console", "w", 2),
+    ];
+    for (filename, mode, file) in args {
+        match freopen(filename, mode, file) {
+            Ok(()) => {}
+            Err(e) => eprintln!("Unable to open {filename} ({e})"),
+        }
+    }
+}
+
+fn boot() {
+    init_rootfs();
+    init_console();
+
+    // TODO: should a failure loading AWS components continue? AWS components are only loaded when
+    // building with the AWS target enabled, so by non-AWS usage this component should not be
+    // loaded.
+    #[cfg(feature = "platform-aws")]
+    match init_platform() {
+        Ok(_) => println!("Successfully sent Nitro heartbeat and loaded necessary kernel modules"),
+        Err(e) => eprintln!("Error when initializing AWS functionality: {e}"),
+    }
+    #[cfg(feature = "platform-aws")]
+    match seed_entropy(4096, get_entropy) {
+        Ok(size) => println!("Seeded kernel with entropy: {size}"),
+        Err(e) => eprintln!("Unable to seed kernel with entropy: {e}"),
+    };
```

Comment:
```suggestion
    #[cfg(feature = "platform-aws")]
    {
        match init_platform() {
            Ok(_) => println!("Successfully sent Nitro heartbeat and loaded necessary kernel modules"),
            Err(e) => eprintln!("Error when initializing AWS functionality: {e}"),
        }
        match seed_entropy(4096, get_entropy) {
            Ok(size) => println!("Seeded kernel with entropy: {size}"),
            Err(e) => eprintln!("Unable to seed kernel with entropy: {e}"),
        };
    }
```
we should be able to feature gate a single code block - which is a little easier to read imo

Reply:
I think the bigger issue here is that seeding entropy only happens on AWS. @lrvick should we pursue an option for seeding entropy on generic targets or is that something handled by the kernel by default?

Reply:
fair enough - thats beyond my scope of knowledge. Just was purely a syntax suggestion :)

Reply:
it's a completely valid point and if it weren't for the fact that the code needed to be split up (which it now is) then it would have been done that way